### PR TITLE
[1.13.x] Make docker stack deploy a little bit more idempotent

### PR DESCRIPTION
**Pull request by vdemeester**
Make sure running multiple times `docker stack deploy --compose-file docker-compose.yml foo` on the same unchanged file doesn't update the services each times. This is mainly due to the use of `map` .

This is related to #29676

This is sorting `environment`, `extraHost` and `port` too. <del>There is some weirdiness with ports but a PR might follow for that.</del> You can reproduce and validate this using the following compose file.

```yaml
version: "3.0"

services:
  test:
    image: busybox
    command: top
    extra_hosts:
      - foo:1.1.1.1
      - bar:2.2.2.2
    environment:
      - FOO=BAR
      - FAR=BOO
    ports:
      - "5000:80"
      - "5001:81"
    networks: [front-tier, back-tier]

networks:
  front-tier:
  back-tier:
```

This also helps with `Error response from daemon: rpc error: code = 2 desc = changing network in service is not supported` when you didn't change the networks.

/cc @thaJeztah @dnephin @icecrime @cpuguy83 @aaronlehmann @aanand 

If it gets in, I'll do a PR against master.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by dnephin** [comment]
It seems to me like these lists should be sorted on the server side before doing the comparison. There could be other clients that hit the API which would benefit from making the change on the server side.

**Comment by vdemeester** [comment]
@dnephin yep, I thought about that after… I can do that

**Comment by vdemeester** [comment]
Server-side alternative is here : https://github.com/docker/docker/compare/1.13.x...vdemeester:29676-deploy-indempotent-swarm-server-side

**Comment by aaronlehmann** [comment]
As explained in https://github.com/docker/docker/pull/29749#issuecomment-269454008, I think this is best done on the client side, to avoid manipulating data inside `Spec` on the server, and so we keep the behavior of the engine API and gRPC API consistent.

**Comment by LK4D4** [comment]
CI fails

**Comment by vdemeester** [comment]
Rebased, should not fail anymore 👼

**Comment by dnephin** [comment]
I don't see sorting as a client-side concern. Is there ever a case where the order of these lists is significant? My understanding is that the order is never significant, so to properly compare the two they need to be sorted to ignore any changes in order.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
ping @dnephin @aaronlehmann PTAL

**Comment by dnephin** [comment]
I still think #29749 was the correct solution.

File: cli/command/stack/deploy.go [hunk 1]
```diff
@@ -611,6 +620,10 @@ func convertService(
 
 	return serviceSpec, nil
 }
+func sortStrings(strs []string) []string {
+	sort.Strings(strs)
+	return strs
+}
```
Comment (aaronlehmann) [inline cli/command/stack/deploy.go]: I don't really like this function. It hides the side effect from `sort.Strings()`. I think it's better to call `sort.Strings()` directly.

**Comment by aaronlehmann** [comment]
LGTM

I don't think there's any harm in doing this. To @dnephin's point, it might also make sense to fix swarmkit internals to be insensitive to ordering of these fields, since it really shouldn't be throwing errors when all you're doing is changing the ordering. These solutions are not mutually exclusive. What we must not do is change the contents of any fields inside a `Spec`, since these are meant to be under user control.

**Comment by vieux** [comment]
ok I'm going to merge this one after https://github.com/docker/docker/pull/30452 is merged, since https://github.com/docker/docker/pull/30452 was a hell to rebase/cherry-pick

**Comment by vieux** [comment]
@vdemeester I rebased

**Comment by thaJeztah** [comment]
All green, merging

**Comment by thaJeztah** [comment]
@vdemeester this needs to be cherry-picked back to master I think; do you open a PR for  that?

**Comment by vdemeester** [comment]
I will tomorrow 😇

**Comment by thaJeztah** [comment]
Also merged in master through https://github.com/docker/docker/issues/30506

**Comment by hairyhenderson** [comment]
> This also helps with `Error response from daemon: rpc error: code = 2 desc = changing network in service is not supported` when you didn't change the networks.

I'm still seeing this error with 1.13.1-rc1, when I didn't change networks. Curiously, only 2 services in my stack (of 9 services) seem to run into this, and these are the 2 that are attached to an `external` network.

This is with client 1.13.1-rc1, but server still at 1.13.0 (Docker for AWS). I'm _assuming_ upgrading the client is enough because this PR was client-only (right!?).

**Comment by aaronlehmann** [comment]
You may have to recreate the services. This PR causes the client to sort the networks to ensure a consistent ordering, but if that ordering doesn't match what was already in the service definition, it won't help avoid the problem...

**Comment by hairyhenderson** [comment]
@aaronlehmann ok, I wondered that... I did end up doing a `docker stack rm` and re-deployed. So far I haven't seen that error again. Thank you! 😁

**Pull request merged**